# Exercise 4.9 Task Part 1

## 3. Import your analysis libraries, as well as your new customer data set as a dataframe.

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [2]:
# Turn project folder path into a string
path = r'C:\Users\Admin\Desktop\Data Analysis\Python\04-2024 Instacart Basket Analysis'

In [3]:
# Checking the path
path

'C:\\Users\\Admin\\Desktop\\Data Analysis\\Python\\04-2024 Instacart Basket Analysis'

In [4]:
# Import customer.csv
customer = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))

In [5]:
# Checking the shape of customer
customer.shape

(206209, 10)

In [6]:
# Checking the output of customer
customer.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [7]:
# Checking data statistics for customer.csv
customer.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [8]:
# Checking columns info
customer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


## 4. Wrangle the data so that it follows consistent logic; for example, rename columns with illogical names and drop columns that don’t add anything to your analysis.

In [9]:
# To rename columns to follow naming convention of all lowercase with _ used for space

In [10]:
# Rename 'First Name' as 'first_name'
customer.rename(columns = {'First Name' : 'first_name'}, inplace = True)

In [11]:
# Rename 'Surnam' as 'last_name'
customer.rename(columns = {'Surnam' : 'last_name'}, inplace = True)

In [12]:
# Rename 'Gender' as 'gender'
customer.rename(columns = {'Gender' : 'gender'}, inplace = True)

In [13]:
# Rename 'STATE' as 'state'
customer.rename(columns = {'STATE' : 'state'}, inplace = True)

In [14]:
# Rename 'Age' as 'age'
customer.rename(columns = {'Age' : 'age'}, inplace = True)

In [15]:
# Rename 'n_dependants' as 'no_of_dependants'
customer.rename(columns = {'n_dependants' : 'no_of_dependants'}, inplace = True)

In [16]:
# Rename 'fam_status' as 'marital_status'
customer.rename(columns = {'fam_status' : 'marital_status'}, inplace = True)

In [17]:
# Check if all columns are renamed properly
customer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   user_id           206209 non-null  int64 
 1   first_name        194950 non-null  object
 2   last_name         206209 non-null  object
 3   gender            206209 non-null  object
 4   state             206209 non-null  object
 5   age               206209 non-null  int64 
 6   date_joined       206209 non-null  object
 7   no_of_dependants  206209 non-null  int64 
 8   marital_status    206209 non-null  object
 9   income            206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


### Columns have been renamed to follow a standard naming convention, no columns is dropped at this point as we do not know which columns is not useful yet.

## 5. Complete the fundamental data quality and consistency checks you’ve learned throughout this Achievement; for example, check for and address missing values and duplicates, and convert any mixed-type data.

In [18]:
# Checking for missing values
customer.isnull().sum()

user_id                 0
first_name          11259
last_name               0
gender                  0
state                   0
age                     0
date_joined             0
no_of_dependants        0
marital_status          0
income                  0
dtype: int64

### 11259 missing value is found on the first name, however, since the last name and user_id do no contain any missing value, it will still allow us to identify the customer despite of these missing values. Thus, these 11259 missing value can be ignore for the time being.

In [19]:
# Checking for duplicate
df_dup = customer[customer.duplicated()]

In [20]:
df_dup

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, no_of_dependants, marital_status, income]
Index: []

### No duplicates found.

In [21]:
# Checking for mixed-type columns
for col in customer.columns.tolist():
  weird = (customer[[col]].map(type) != customer[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (customer[weird]) > 0:
    print (col)

first_name


### first_name column identified as mixed-type columns.

In [22]:
# Convert 'first_name' column to string type column
customer['first_name'] = customer['first_name'].astype('str')

In [23]:
# Checking for mixed-type columns again after conversion
for col in customer.columns.tolist():
  weird = (customer[[col]].map(type) != customer[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (customer[weird]) > 0:
    print (col)

### Mixed-type columns have been resolved, current columns do not contain anymore mixed-type columns.

## 6. Combine your customer data with the rest of your prepared Instacart data. (Hint: Make sure the key columns are the same data type!)

In [24]:
# Import orders_products_aggregated.pkl
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_aggregated.pkl'))

In [25]:
# Check shape of ords_prods_merge
ords_prods_merge.shape

(32404289, 22)

In [26]:
# Check output of ords_prods_merge
ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  prices     busiest_day  busiest_day_new  \
0         Soda  ...     9.0  Regularly busy   Regularly busy   
1         Soda  ...     9.0  Regularly busy     Slowest days   
2         Soda  ...     9.0  Regularly busy     Slowest days   
3         Soda  ...     9.0      Least busy     Slowest days   
4         Soda  ...     9.0      Least busy     Slowest days   

  busiest_period_of_day max_order  loyalty_flag  avg_spending Spender_status  \
0        Average orders        10  New customer      6.367797    Low spender   
1        Average orders        10  New customer      6.367797    Low spender   
2           Most orders        10  New customer      6.367797    Low spender   
3        Average orders        10  New customer      6.367797    Low spender   
4           Most orders        10  New customer      6.367797    Low spender   

   median_order_days        order_frequency  
0               20.5  Non-frequent customer  
1               20.5  Non-frequent customer  
2               20.5  Non-frequent customer  
3               20.5  Non-frequent customer  
4               20.5  Non-frequent customer  

[5 rows x 22 columns]

In [27]:
# Check data type of ords_prods_merge columns
ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404289 entries, 0 to 32404288
Data columns (total 22 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   order_number            int64  
 3   order_day_of_week       int64  
 4   order_hour_of_day       int64  
 5   days_since_prior_order  float64
 6   product_id              int64  
 7   add_to_cart_order       int64  
 8   reordered               int64  
 9   product_name            object 
 10  aisle_id                int64  
 11  department_id           int64  
 12  prices                  float64
 13  busiest_day             object 
 14  busiest_day_new         object 
 15  busiest_period_of_day   object 
 16  max_order               int64  
 17  loyalty_flag            object 
 18  avg_spending            float64
 19  Spender_status          object 
 20  median_order_days       float64
 21  order_frequency         objec

In [28]:
# Merge ords_prods_merge with customer data with 'user_id' column
ords_prods_customer_merge = ords_prods_merge.merge(customer, on = 'user_id', how = 'left', indicator = True)

In [29]:
# Checking shape of ords_prods_customer_merge
ords_prods_customer_merge.shape

(32404289, 32)

In [30]:
# Checking output of ords_prods_customer_merge
ords_prods_customer_merge.head(10)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   
5   3367565        1             6                  2                  7   
6    550135        1             7                  1                  9   
7   3108588        1             8                  1                 14   
8   2295261        1             9                  1                 16   
9   2550362        1            10                  4                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   
5                    19.0         196                  1          1   
6                    20.0         196                  1          1   
7                    14.0         196                  2          1   
8                     0.0         196                  4          1   
9                    30.0         196                  1          1   

  product_name  ...  first_name  last_name  gender    state age date_joined  \
0         Soda  ...       Linda     Nguyen  Female  Alabama  31   2/17/2019   
1         Soda  ...       Linda     Nguyen  Female  Alabama  31   2/17/2019   
2         Soda  ...       Linda     Nguyen  Female  Alabama  31   2/17/2019   
3         Soda  ...       Linda     Nguyen  Female  Alabama  31   2/17/2019   
4         Soda  ...       Linda     Nguyen  Female  Alabama  31   2/17/2019   
5         Soda  ...       Linda     Nguyen  Female  Alabama  31   2/17/2019   
6         Soda  ...       Linda     Nguyen  Female  Alabama  31   2/17/2019   
7         Soda  ...       Linda     Nguyen  Female  Alabama  31   2/17/2019   
8         Soda  ...       Linda     Nguyen  Female  Alabama  31   2/17/2019   
9         Soda  ...       Linda     Nguyen  Female  Alabama  31   2/17/2019   

   no_of_dependants marital_status  income _merge  
0                 3        married   40423   both  
1                 3        married   40423   both  
2                 3        married   40423   both  
3                 3        married   40423   both  
4                 3        married   40423   both  
5                 3        married   40423   both  
6                 3        married   40423   both  
7                 3        married   40423   both  
8                 3        married   40423   both  
9                 3        married   40423   both  

[10 rows x 32 columns]

In [31]:
# checking result of merge using merge flag
ords_prods_customer_merge['_merge'].value_counts()

_merge
both          32404289
left_only            0
right_only           0
Name: count, dtype: int64

In [33]:
ords_prods_merge['prices'].value_counts()

prices
12.3    624261
10.3    510496
4.0     447272
6.3     438467
1.3     425691
         ...  
20.2       123
22.7        99
21.5        85
18.3        11
21.0         8
Name: count, Length: 242, dtype: int64

### Both dataset have been successfully merged.

## 8. Export this new dataframe as a pickle file so you can continue to use it in the second part of this task.

In [32]:
# Exporting ords_prods_customer_merge as pkl
ords_prods_customer_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_customer_merge.pkl'))